File: proposed-projects/round-trip-experiment/reverse/20260908T234247Z_2da5a2c0/20260910T040422Z_0f628863/ported_implementation/src/grammar-graph/print.ts
import { Color, type Colorize } from './colorize.ts';
import { getParentStackId } from './get-parent-stack-id.ts';
import type { GraphNode } from './graph-node.ts';
import type { GraphPointer } from './graph-pointer.ts';
import { isRange, isRuleChar, isRuleRef } from './type-guards.ts';

export const printGraphPointer = (pointer: GraphPointer, colorize: Colorize): string =>
  colorize(`*${getParentStackId(pointer, colorize)}`, Color.RED);

export const printGraphNode = (
  node: GraphNode,
  {
    pointers,
    showPosition = false,
    colorize: col,
  }: {
    pointers?: Iterable<GraphPointer>;
    showPosition?: boolean;
    colorize: Colorize;
  }
): string => {
  const rule = node.rule;

  const parts: string[] = [];
  if (showPosition) {
    parts.push(col('{', Color.BLUE));
    parts.push(col(node.id, Color.GRAY));
    parts.push(col('}', Color.BLUE));
  }
  if (isRuleChar(rule)) {
    const values: string[] = [];
    for (const value of rule.value) {
      if (isRange(value)) {
        // The inner array stringifies with commas between its entries.
        values.push(value.map((val) => col(String.fromCodePoint(val), Color.YELLOW)).join(','));
      } else {
        values.push(getChar(value));
      }
    }
    parts.push(col('[', Color.GRAY));
    parts.push(col(values.join(''), Color.YELLOW));
    parts.push(col(']', Color.GRAY));
  } else if (isRuleRef(rule)) {
    parts.push(
      col('Ref(', Color.GRAY) + col(`${rule.value}`, Color.GREEN) + col(')', Color.GRAY)
    );
  } else {
    parts.push(col(rule.type, Color.YELLOW));
  }

  if (pointers) {
    for (const pointer of pointers) {
      const pointerParts: string[] = [];
      if (pointer.node === node) {
        pointerParts.push(pointer.print({ colorize: col }));
      }
      if (pointerParts.length) {
        parts.push(col('[', Color.GRAY));
        parts.push(...pointerParts);
        parts.push(col(']', Color.GRAY));
      }
    }
  }

  const nextPrinted =
    node.next !== undefined
      ? printGraphNode(node.next, { pointers, colorize: col, showPosition })
      : undefined;

  return [parts.join(''), nextPrinted].filter(Boolean).join(col('-> ', Color.GRAY));
};

const getChar = (charCode: number): string => {
  const char = String.fromCodePoint(charCode);
  if (char === '\n') {
    return '\\n';
  }
  return char;
};

File: packages/gbnf-experiment/data/20260909T193332Z_f85fdfbb/ported_implementation/grammar_graph/print.py
from __future__ import annotations

from .colorize import Color, colorize as default_colorize
from .get_parent_stack_id import get_parent_stack_id
from .type_guards import is_range, is_rule_char, is_rule_ref


def print_graph_pointer(pointer, colorize=default_colorize) -> str:
    col = colorize
    return col(f"*{get_parent_stack_id(pointer, col)}", Color.RED)


def _get_char(char_code: int) -> str:
    char = chr(char_code)
    if char == "\n":
        return "\\n"
    return char


def print_graph_node(
    node,
    pointers=None,
    show_position: bool = False,
    colorize=default_colorize,
) -> str:
    col = colorize
    rule = node.rule

    parts: list[str] = []
    if show_position:
        parts.append(col("{", Color.BLUE))
        parts.append(col(node.id, Color.GRAY))
        parts.append(col("}", Color.BLUE))

    if is_rule_char(rule):
        rendered = []
        for v in rule.value:
            if is_range(v):
                # A nested array stringifies with commas between its members.
                rendered.append(",".join(col(chr(val), Color.YELLOW) for val in v))
            else:
                rendered.append(_get_char(v))
        parts.append(col("[", Color.GRAY))
        parts.append(col("".join(rendered), Color.YELLOW))
        parts.append(col("]", Color.GRAY))
    elif is_rule_ref(rule):
        parts.append(
            col("Ref(", Color.GRAY)
            + col(f"{rule.value}", Color.GREEN)
            + col(")", Color.GRAY)
        )
    else:
        parts.append(col(str(rule.type), Color.YELLOW))

    if pointers:
        for pointer in pointers:
            pointer_parts: list[str] = []
            if pointer.node is node:
                pointer_parts.append(pointer.print(colorize=col))
            if pointer_parts:
                parts.append(col("[", Color.GRAY))
                parts.extend(pointer_parts)
                parts.append(col("]", Color.GRAY))

    tail = (
        node.next.print(pointers=pointers, colorize=col, show_position=show_position)
        if node.next is not None
        else None
    )
    return col("-> ", Color.GRAY).join(p for p in ["".join(parts), tail] if p)
